# 351 Midterm
### Kaetlyn Gibson

---

# Computer Problems

### Problem 1
Consider the approximation

\begin{align}
\frac{d}{dx}tan x ≈\frac{tan(x + h)−tan(x)}{h}
\end{align}

where *h = 2<sup>−n</sup>* and *n* is a positive integer. If we’re using floating point arithmetic,
find (approximately) the value of *n* for which this approximation works well for values
of *x* in the interval *[−π/4, π/4]*. (If *h* is too small, there’s loss of significance in the
computation, whereas if *h* is too big, the approximation is just bad.)


In [437]:
from numpy import *

def p1(n, x):
    h = 2**(-n)
    approx = (tan(x+h)-tan(x))/h
    actual = (1/cos(x))**2 # sec^2(x) = 1
    errorCheck = abs(actual-approx)
    return errorCheck
    
def testValues(potentialNs, intervalValues):
    for n in potentialNs:
        print(f"testing n={n} among some interval values")
        for val in intervalValues:
            print(f"==>interval value = {val}")
            result = p1(n, val)
            print(f"====> error abs(derivative of tanx - approximation): {result}")
        print("")

potentialNs = [100, 50, 31, 30, 29, 25, 1, 0]
intervalValues = [(-pi)/4,(-pi/6), -0.1, 0.1, (pi/6), (-pi)/4]

testValues(potentialNs, intervalValues)

testing n=100 among some interval values
==>interval value = -0.7853981633974483
====> error abs(derivative of tanx - approximation): 1.9999999999999996
==>interval value = -0.5235987755982988
====> error abs(derivative of tanx - approximation): 1.3333333333333333
==>interval value = -0.1
====> error abs(derivative of tanx - approximation): 1.0100670464224946
==>interval value = 0.1
====> error abs(derivative of tanx - approximation): 1.0100670464224946
==>interval value = 0.5235987755982988
====> error abs(derivative of tanx - approximation): 1.3333333333333333
==>interval value = -0.7853981633974483
====> error abs(derivative of tanx - approximation): 1.9999999999999996

testing n=50 among some interval values
==>interval value = -0.7853981633974483
====> error abs(derivative of tanx - approximation): 4.440892098500626e-16
==>interval value = -0.5235987755982988
====> error abs(derivative of tanx - approximation): 0.04166666666666674
==>interval value = -0.1
====> error abs(derivativ

My thinking: calculate the approximation for the derivative of tan(x) given above and compare it to the derivative of tan(x), sec<sup>2</sup>(x) (by using 1/cos<sup>2</sup>(x) since there isn't sec in numpy, and sec(x) = 1/cos(x)). I then calculate error by subtracting the approximation from the "actual" value. I do this for multiple values of n and for multiple values x in the given interval *[−π/4, π/4]*.

My results: based on some testing(above), n=30 works very well for values of x in the interval *[−π/4, π/4]*, as the error is quite small for all x values tested compared to other n values. 

### Problem 2 (Ignore; I don't think I did this right, but I'll leave it in case you might have some input on if I'm even on the right track.)
Consider the function 
\begin{align}
f(x) =
x^7 + 105x^6 + 4725x^5 + 118125x^4 + 1771875x^3 + 15946875x^2 + 79734374x + 170859367.
\end{align}

There’s a particularly nice value *c* for which the Taylor series for *f* centered at the
point *x = c* has only three nonzero coefficients. What is this *c*?


In [187]:
#complete horners
def horner(x, a, popped):
    n = len(a)-1
    rb = [a[-1]]
    b = [a[-1]]
    # ... continue.  You can now use a[0], ..., a[n] for the coefficients of p.
    for i in range(n-1, -1, -1):
        rb.append(a[i] + x*b[-1])
        b.append(rb[-1])
    r = b.pop()
    popped.append(r) # pop p(x) off the list
    return (b, popped)

def repeat(x, a, popped):
    result = horner(x, a, popped)
    while len(result[0]) > 0: # while still stuff to pop
        newA = result[0][::-1] # reverse this since output of horner function is in reverse of input
        newPopped = result[1] # keep track of popped
        result = horner(x, newA, newPopped) # call again with new values
    return result[1] 
    
def testValues(a):
    for i in range(-100000, 100000):
        count = 0
        result = repeat(i, a, [])
        # if there is a nonzero coefficient, count it
        for coeff in result:
            if coeff != 0:
                count += 1
        # if there are 3 nonzero coefficients, we found the right c (doesn't print otherwise)
        if count == 3:
            print(result)
            print(i)

    
def testValuesFrac(a):
    for i in range(-1000, 1000):
        count = 0
        if i != 0:
            result = repeat(1/i, a, [])
            # if there is a nonzero coefficient, count it
            for coeff in result:
                if coeff != 0:
                    count += 1
            # if there are 3 nonzero coefficients, we found the right c (doesn't print otherwise)
            if count == 3:
                print(result)
                print(i)

# check example 3 in textbook on page 22.     
# repeat(2, [-5, -12, 13, 15, -2, 3],[]) 
# Question: What is the Taylor series of the function f (x) = 3x^5 − 2x^4 + 15x^3 + 13x^2 − 12x − 5 at the point c = 2?
# Answer: f(x) = 207 + 396(x − 2) + 295(x − 2)^2 + 119(x − 2)^3 + 28(x − 2)^4 + 3(x − 2)^5
# My result: [207, 396, 295, 119, 28, 3]
# Looks like it works!...


repeat((1/15), [1,105,4725,118125,1771875,15946875,79734374,170859367],[])
# repeat((15), [1,105,4725,118125,1771875,15946875,79734374,170859367],[])
# testValues([1,105,4725,118125,1771875,15946875,79734374,170859367])
# testValuesFrac([1,105,4725,118125,1771875,15946875,79734374,170859367])

[127.99999986538636,
 6719.999987182441,
 151199.9994824691,
 1889999.9885432099,
 14174999.85037037,
 63787498.85333333,
 159468745.26666668,
 170859367]

My thinking: I have a couple of functions that perform the complete Horner's algorithm. I can use this to get the coefficients of the Taylor series of f(x) centered at x=c. I can then test values of c until there are three nonzero coefficients. I'll go through a large range of numbers, and then see where there were only 3 nonzero coefficients.

My results: I am uncertain as of the moment. I do know that 5 of the terms are divisible by 105... and 15, so perhaps something to do with this. But no luck.

### Problem 3 
Implement the secant method for root finding. Use it to find the number that you get
when you push the ”cos” button on a scientific calculator repeatedly.

In [436]:
from math import *

def f(x):
    """ equation for secant method """
#     f = x**5 + x**3 + 3 # from example 1 on page 113
    f = cos(x)
    return f

def secantMethod(a, b, nmax, e):
    """ secant method for root finding as described on page 112-113, chapter 3 of the textbook """
    fa = f(a)
    fb = f(b)
    if abs(fa) > abs(fb):
        temp = a
        tempf = fa
        a = b
        b = temp
        fa = fb
        fb = tempf
    print(f"n: {0}, xn: {a}, p(xn): {fa}")
    print(f"n: {1}, xn: {b}, p(xn): {fb}")
    
    for n in range(2, nmax+1):
        if abs(fa) > abs(fb):
            temp = a
            tempf = fa
            a = b
            b = temp
            fa = fb
            fb = tempf
        d = (b-a)/(fb-fa)
        b = a
        fb = fa
        d = d * fa
        if abs(d) < e:
            print("convergence")
            return
        a = a -d
        fa = f(a)
        print(f"n: {n}, xn: {a}, p(xn): {fa}")


# secantMethod(-1, 1, 8, 0.000000001) # from example 1 on page 113
secantMethod(0, 3, 100, 0.000000001)

n: 0, xn: 3, p(xn): -0.9899924966004454
n: 1, xn: 0, p(xn): 1.0
n: 2, xn: 1.5075433727137042, p(xn): 0.06321078401244082
n: 3, xn: 1.597117109571447, p(xn): -0.026317743780686558
n: 4, xn: 1.57078607616556, p(xn): 1.0250629336342976e-05
n: 5, xn: 1.5707963279781092, p(xn): -1.1832125777453532e-09
n: 6, xn: 1.5707963267948966, p(xn): 6.123233995736766e-17
convergence


I implemented the secant method for root finding. However, I am uncertain as to what the number is that is meant by pushing the "cos" button on a scientific calculator repeatedly. I don't get a number when I do this, only error. Unless that was just a jokingly way of saying to find a root of cos(x)...

### Problem 4 (Ignore; did not do)
Let *f<sub>n</sub>* be the degree-*n* Taylor polynomial for sin(*x*) around the point *x* = 0, and let *e<sub>n</sub>(x) = sin(x) − f<sub>n</sub>(x)*. Let *M(n)* be the smallest number *x* > 0 for which *e<sub>n</sub>(x)* exceeds 0.001. Plot a graph of *M(n)* versus *n*, for *n* < 20.


---

# Written Problems

### Problem 1
Using Taylor’s theorem, find a number n such that the value of the degree-n Taylor
polynomial for ln(x), centered at x = 1, is within 2<sup>−16</sup> of the value of ln(x) itself for
all x in ($\frac{1}{3}$, $\frac{3}{2}$). You don’t need to find the smallest such n - but you do need to provide
reasoning for why the value of n that you chose works.

Let f<sub>n</sub>(x) represent the degree-n Taylor polynomial for ln(x), centered at x=1. (using some stuff from example 9 in ch 1.2)

In [432]:
import math

def func_ln(x, n):
    ln_approx = 0
    for i in range(1,n+1):
        ln_approx += ((-1)**(i-1)) * (((x-1)**i)/i)
    
    return ln_approx

def errorCheck(x):
    for n in range(0,50):
        approx = func_ln(x, n)
        exp = math.log(x)
        error = abs(approx - exp)
        if error < 2**(-16):
            break
    print(f'{n} terms: Taylor Series approx= {approx}, exp calc= {exp}, error = {error}')
    
errorCheck(1.25)
errorCheck(1.2)
errorCheck(0.34)
errorCheck(0.5)
errorCheck(1/3)
errorCheck(3/2)

6 terms: Taylor Series approx= 0.22313639322916667, exp calc= 0.22314355131420976, error = 7.158085043090789e-06
5 terms: Taylor Series approx= 0.18233066666666664, exp calc= 0.1823215567939546, error = 9.109872712048661e-06
21 terms: Taylor Series approx= -1.0787963861607122, exp calc= -1.0788096613719298, error = 1.3275211217633753e-05
13 terms: Taylor Series approx= -0.6931389804314607, exp calc= -0.6931471805599453, error = 8.20012848457008e-06
22 terms: Taylor Series approx= -1.0986015077404407, exp calc= -1.0986122886681098, error = 1.0780927669040707e-05
11 terms: Taylor Series approx= 0.4054790370220058, exp calc= 0.4054651081081644, error = 1.3928913841387836e-05


Take ln($\frac{5}{4}$) for a test. Using the series for ln(x) centered at x=1, f(1.25) = $\frac{1}{4}$ - $\frac{1}{32}$ + $\frac{1}{192}$ - $\frac{1}{1024}$ +...+ $\frac{(-1)^{n-1}}{(4)^n*n}$

By the Alternating Series Thm, the error for truncating the series by n terms is | ln(1.25)-f(1.25) | <= $\frac{1}{(4)^{n+1}(n+1)}$.
We want to select n such that $\frac{1}{(4)^{n+1}(n+1)}$ < 2<sup>-16</sup>. We found n=6 will work, and we successfully check below.

We can check |ln(1.25) - f<sub>6</sub>(1.25)| = (about) 7.16 * 10^-6

and |ln(1.25) - f<sub>5</sub>(1.25)| = (about) 3.35 * 10^-5

n=22 could work for all the values between ($\frac{1}{3}$, $\frac{3}{2}$). I checked how it would work for the ends of the interval (though not included in valid values of x) to see the highest value of n it would take.

### Problem 2
Suppose f(x) is a degree-n polynomial with n distinct roots in the interval [−1, 1].
Describe an algorithm which finds, numerically, all of these roots with an error of at
most $\in$ . Hint: Between any two roots of f, you can find a root of f ' (why?).

Between any two roots of f, you can find a root of f'. This is because as f crosses the x-xis, it is either increasing or decreasing. The root of f' that is between the two roots will reflect this change of either increasing then decreasing at a turning point, or decreasing then increasing at a turning point, as the moment it changes from increasing to decreasing at a turning point(or vice-versa) will be the root of f'. For example, f(x)=x<sup>2</sup>-1 has two roots x=-1, 1 and f'(x)=2x has root x=0. Another example is f(x)=5x<sup>3</sup>+2x<sup>2</sup>-3x (roots at x=-1, 0, 0.6) and f'(x)=15x<sup>2</sup>+4x-3 (roots at x=-0.6, 0.333...).


Knowing this, you could take the integral of f(x), and then see where there are turning points in the antiderivative. These turning points should line up with the roots of f(x).

### Problem 3
Here is one way that a computer might compute ln x, for a fixed number x. Write y = ln x for the number we’re trying to compute. Then e<sup>y</sup> = x, so y is a root of e<sup>y</sup>−x. We find this root with Newton’s method. For which values of x does this work well? For which values does it work badly?


We use the formula for Newton's method, x<sub>n+1</sub> = x<sub>n</sub> - $\frac{f(x_n)}{f'(x_n)}$, and apply it by substituting f(x) = e<sup>y</sup> - *x*

This looks like, x<sub>n+1</sub> = x<sub>n</sub> - $\frac{e^y-x}{e^y}$

---
From first glance, it looks like this will not work for values of x < 0, since ln(x) only works with values x > 0 (else it is undefined).

Let's try x = 0. We can start with x<sub>0</sub> = 0

ln(0) is undefined
x<sub>1</sub> = 0 - $\frac{e^{0}-0}{e^{0}}$ = -1

x<sub>2</sub> = x<sub>1</sub> - $\frac{e^{x_1}-0}{e^{x_1}}$ = -2

x<sub>3</sub> = x<sub>2</sub> - $\frac{e^{x_2}-0}{e^{x_2}}$ = -3

x<sub>4</sub> = x<sub>3</sub> - $\frac{e^{x_3}-0}{e^{x_3}}$ = -4

That is weird. This behaves badly, as expected.

---
Let's try x = -50. We can start with x<sub>0</sub> = 0

ln(-50) is undefined
x<sub>1</sub> = 0 - $\frac{e^{0}+50}{e^{0}}$ = -51

x<sub>2</sub> = x<sub>1</sub> - $\frac{e^{x_1}-0}{e^{x_1}}$ = -7.04674541 * 10<sup>23</sup>

x<sub>3</sub> = x<sub>2</sub> - $\frac{e^{x_2}-0}{e^{x_2}}$ = ? I got a divide by zero error

That is weird. This also behaves badly, as expected.

---
Let's try x = 2. We can start with x<sub>0</sub> = 0.5

ln(2) is approximately 0.6931471806
x<sub>1</sub> = 0.5 - $\frac{e^{0.5}-2}{e^{0.5}}$ = 0.7130613194 (about)

x<sub>2</sub> = x<sub>1</sub> - $\frac{e^{x_1}-2}{e^{x_1}}$ = 0.6933441573 (about)

x<sub>3</sub> = x<sub>2</sub> - $\frac{e^{x_2}-2}{e^{x_2}}$ = 0.6931472

x<sub>4</sub> = x<sub>3</sub> - $\frac{e^{x_3}-2}{e^{x_3}}$ = 0.6931471806 (about)

That was quick!

---
Let's try x = 50. We can start with x<sub>0</sub> = 3

ln(50) is approximately 3.912023005
x<sub>1</sub> = 3 - $\frac{e^{3}-50}{e^{3}}$ = 4.489353418 (about)

x<sub>2</sub> = x<sub>1</sub> - $\frac{e^{x_1}-50}{e^{x_1}}$ = 4.05074879 (about)

x<sub>3</sub> = x<sub>2</sub> - $\frac{e^{x_2}-50}{e^{x_2}}$ = 3.921215441 (about)

x<sub>4</sub> = x<sub>3</sub> - $\frac{e^{x_3}-50}{e^{x_3}}$ = 3.912065127 (about)

x<sub>5</sub> = x<sub>4</sub> - $\frac{e^{x_4}-50}{e^{x_4}}$ = 3.912023006 (about)

x<sub>6</sub> = x<sub>5</sub> - $\frac{e^{x_5}-50}{e^{x_4}}$ = 3.912023005 (about)

That was also pretty quick!

Based on my observations above, I think as long as x > 0, then Newton's Method should behave well. Values x <= 0 behaved very badly.

### Problem 4
Use Newton’s method to compute $\sqrt[3]{a}$, where *a* is a fixed number. Make sure to simplify
the expression so as to avoid loss of significance.

We are trying to find the cubed root of *a*, a fixed number. Let's say that x is the cubed root of *a*, therefore x<sup>3</sup> = *a*.

We set this equal to zero: x<sup>3</sup> - *a* = 0.

We then use the formula for Newton's method, x<sub>n+1</sub> = x<sub>n</sub> - $\frac{f(x_n)}{f'(x_n)}$, and apply it by substituting f(x) = x<sup>3</sup> - *a*

This looks like: x<sub>n+1</sub> = x<sub>n</sub> - $\frac{(x_n)^3-a}{3*(x_n)^2}$